tests/dofs/dof_tools_12a.cc

```diff
@@ -0,0 +1,133 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2003 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+#include <deal.II/base/logstream.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/mapping_q_generic.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/tria_iterator.h>
+
+#include <algorithm>
+#include <fstream>
+#include <iomanip>
+#include <string>
+
+#include "../tests.h"
+
+// check
+//   DoFTools::extract_dofs as in dof_tools_12 in parallel for fewer elements
+
+void
+output_bool_vector(std::vector<bool> &v)
+{
+  for (unsigned int i = 0; i < v.size(); ++i)
+    deallog << (v[i] ? '1' : '0');
+  deallog << std::endl;
+}
+
+
+template <int dim, typename DoFHandlerType>
+void
+check_this(const DoFHandlerType &dof_handler)
+{
+  IndexSet locally_relevant_dofs;
+  DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+
+  std::vector<bool> selected_dofs(dof_handler.n_locally_owned_dofs());
+  std::vector<bool> mask(dof_handler.get_fe().n_components(), false);
+
+  // only select first component
+  mask[0] = true;
+  DoFTools::extract_dofs(dof_handler, ComponentMask(mask), selected_dofs);
+  if (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+    {
+      output_bool_vector(selected_dofs);
+    }
+
+  // also select last component
+  mask.back() = true;
```

Comment:
Currently, you are only ever checking `FE_Q` elements, i.e. scalar ones with only one component. Thus, the size of `mask` is one and you are testing the same thing twice.

Reply:
Thanks for the review! You are absolutely right, I added vector valued elements and (most) of the elements of dof_tools_common.
I also use now "MPILogInitAll" to get the output of all processes, but, as you said, the selected_dofs vectors are just the locally owned vectors, which are not mapped to global DoF indices. I don't think that this is a problem in this testing scenario, or am I wrong there? Therefore I wouldn't change this, if there is no objection.

Greetings
Mathias

Reply:
No, it's not a problem that you are considering the locally owned degrees of freedom. At least for me, it is just unintuitive that the vector returned doesn't correspond to actual degrees of freedoms but has to be mapped using the `IndexSet` of locally owned DoFs to make sense (with a distributed mesh). We often define a separate function returning an `IndexSet` instead of a `std::vector<bool>` when porting functions from serial `Triangulations` to distributed ones. Of course, this behavior was pre-existing for the non-hp version and it's fine to extend that behavior to the hp case. Nevertheless, I would appreciate a comment in the documentation on how to interpret the returned object.

Reply:
masterleinad: OK, I understand. I explained this fact in the comment of the beginning of the test, or is there another place, where this should go?

Reply:
Sorry for not being precise. Such a comment should go to https://github.com/dealii/dealii/blob/6e6a2bf22ba1489ad399aaa06e5d9f6431fcb7fb/include/deal.II/dofs/dof_tools.h#L1318-L1350

Reply:
Ah, I didn't understand this, it was just "obvious" to me, that this is, what the function does. I have added a little note in the description and hope this is OK.

Reply:
Yes, that looks OK. How did you verify the correctness of the test output? :wink:

Reply:
I didn't... correct until someone else proofs the opposite ;)